feat: Dedicated release note for dropping repository permissions for Administration on GitHub (#1971)
codacy/docs: docs/release-notes/cloud/cloud-2024-01-DD-gh-repository-ssh-keys-deprecation.md

---
rss_title: Codacy release notes RSS feed
rss_href: /feed_rss_created.xml
---

> # Deprecation of SSH keys for GitHub repositories January DD, 2024<!--TODO PLUTO-802 Update release date-->

Author: I reviewed the content using "discontinuation", and "stop using" later.

!!! info "This release note applies only to GitHub"

> For increased security, Codacy deprecated the usage of **repository SSH keys** for Git operations on GitHub in favor of **installation access tokens**.

Author: You're right, "deprecation" might not exactly fit this change and can be confusing. I tweaked the terminology using "discontinuation".

This change translates into important security improvements:

> -   It reduces the liability of the Codacy GitHub App, as it no longer requires [read and write repository permissions for Administration](https://docs.github.com/en/rest/authentication/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-administration).

Reviewer: Can we reformulate this in the positive?

```suggestion
-   It improves the security of the Codacy GitHub App, as it no longer requires [read and write repository permissions for Administration](https://docs.github.com/en/rest/authentication/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-administration).
```

Author: I see your point... However, I would like to use here a word related to "liability", and I'm not finding a good positive expression to replace it.
"It improves the security" kind of repeats the stem sentence "This change translates into important **security improvements**"... 🤔 

I'll keep the current expression for now. 🙏

> -   It reduces the liability of the Codacy GitHub App, as it no longer requires [read and write repository permissions for Administration](https://docs.github.com/en/rest/authentication/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-administration).

Author: I reviewed this paragraph. 👍

-   Contrary to long-living SSH keys, [installation access tokens](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-an-installation-access-token-for-a-github-app#about-installation-access-tokens) expire after one hour and Codacy loses access when the GitHub App is uninstalled.

To minimize any impact, Codacy deprecated the usage of repository SSH keys for now but will keep using them as a fallback mechanism at this first stage. At a later stage, Codacy will [delete the repository SSH keys](#timeline) stored in our systems.

## Changes on the Codacy GitHub App required permissions {: id="changes-permissions"}

To use installation access tokens, the Codacy GitHub App now requires [repository read permissions for Contents](https://docs.github.com/en/rest/authentication/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-contents).

> On the other hand, the Codacy GitHub App no longer requires [read and write repository permissions for Administration](https://docs.github.com/en/rest/authentication/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-administration). This permission will be [removed from the Codacy GitHub App](#timeline) at a later stage.

Reviewer 2: "no longer requires" is present tense. The app will only drop this requirement on the target date of February 19. Unfortunately we cannot fine tune the app to drop this permission for people who have already accepted `Contents` permission. Once we remove the permission from the App (February 19) it will automatically cascade to all installations and be dropped.

Author: Thanks for the input. I reviewed this paragraph. 👍

## What do you need to do if you have a GitHub organization? {: id="to-do"}

Make sure an organization owner [approves the updated permissions for the Codacy GitHub App](https://docs.github.com/en/apps/using-github-apps/reviewing-and-modifying-installed-github-apps) on your GitHub organization, if not done yet.

> Codacy has been requesting repository read permissions for **Contents** since September 2023. Organization owners should had received a GitHub notification to review a request for this additional permission:

Reviewer: Typo

```suggestion
Codacy has been requesting repository read permissions for **Contents** since September 2023. Organization owners should have received a GitHub notification to review a request for this additional permission:
```

![Codacy GitHub App updated permissions request](../images/2023-12-DD-gh-updated-permissions.png)<!--TODO PLUTO-802 Rename image-->

If you have any questions or need help, please contact <mailto:support@codacy.com>.

## Removal of repository permissions for Administration and SSH keys {: id="timeline"}

To ensure the conditions to use installation access tokens on GitHub organizations are met before the removal of repository permissions for Administration and SSH keys, Codacy will execute a phased rollout according to the timeline below:

<table>
  <thead>
    <th>Date</th>
    <th>Event</th>
    <th>What to expect</th>
  </thead>
  <tbody>
    <tr>
      <td>January DD, 2024</td><!--TODO PLUTO-802 Update release date-->

>       <td>Codacy started using installation access tokens to clone and integrate with your repositories, and deprecated the usage of repositories SSH keys</td>
>       <td>
>         <p>From this day on, repositories SSH keys are used exclusively as a fallback mechanism when the Contents permission is missing.</p>

Reviewer: I'd double-check, but I believe these noun modifiers [should be singular](https://teflpedia.com/Modifier_noun#:~:text=In%20English%2C%20the%20modifier%20noun,meaning%20of%20the%20head%20noun.):

```suggestion
      <td>Codacy started using installation access tokens to clone and integrate with your repositories, and deprecated the usage of repository SSH keys</td>
      <td>
        <p>From this day on, repository SSH keys are used exclusively as a fallback mechanism when the Contents permission is missing.</p>
```

Author: You're right! 🤓

      </td>
    </tr>
    <tr>
      <td>February 19, 2024</td>
      <td>Codacy will remove repository permissions for Administration from the Codacy GitHub App and drop the usage of repository SSH keys</td>
      <td>
        <p>Codacy will start using installation access tokens to clone and integrate with your repositories, with no fallback mechanisms. Also, repository permissions for Administration will be removed from the Codacy GitHub App.</p>
        <p>If on this date the Codacy GitHub App updated permissions haven't been approved on your GitHub organization yet, your Codacy quality analysis will fail and you'll not be able to add new repositories to Codacy until an organization owner <a href="#to-do">approves the updated permissions.</a></p>
      </td>
    </tr>
    <tr>
      <td>To define</td>
      <td>Codacy will no longer store repository SSH keys</td>
      <td>

>         <p>On this day, Codacy will delete all the repository SSH keys stored in our systems.</p>

Reviewer 3: Maybe we can just mention that users can delete Deploy keys themselves in GH 
https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/reviewing-your-deploy-keys

for context link to your repo (I hope you have access): 
https://github.com/troubleshoot-codacy/test-app-analysis/settings/keys

Author: Yes, that is important information to mention. 🙏 
I added it, with a link to GH documentation.
